python/test/unit/fem/test_assembler.py

```diff
@@ -1560,3 +1560,159 @@ def test_mixed_quadrature(dtype, method):
+@pytest.mark.parametrize("dtype", [np.float32, np.float64, np.complex64, np.complex128])
+def test_vertex_integral_rank_1(cell_type, ghost_mode, dtype):
+    comm = MPI.COMM_WORLD
+    rdtype = np.real(dtype(0)).dtype
+
+    msh = None
+    cell_dim = mesh.cell_dim(cell_type)
+    if cell_dim == 1:
+        msh = mesh.create_unit_interval(comm, 4, ghost_mode=ghost_mode, dtype=rdtype)
+    elif cell_dim == 2:
+        msh = mesh.create_unit_square(
+            comm, 4, 4, cell_type=cell_type, ghost_mode=ghost_mode, dtype=rdtype
+        )
+    elif cell_dim == 3:
+        msh = mesh.create_unit_cube(
+            comm, 4, 4, 4, cell_type=cell_type, ghost_mode=ghost_mode, dtype=rdtype
+        )
+
+    vertex_map = msh.topology.index_map(0)
+    num_vertices = vertex_map.size_local
+
+    def check_vertex_integral_against_sum(form, vertices, weighted=False):
+        weights = vertex_map.local_to_global(vertices) if weighted else np.ones_like(vertices)
+        expected_value_l = np.zeros(num_vertices, dtype=rdtype)
+        expected_value_l[vertices] = msh.geometry.x[vertices, 0] * weights
+        value_l = fem.assemble_vector(fem.form(form, dtype=dtype))
+        equal_l = np.allclose(
+            expected_value_l, np.real(value_l.array[:num_vertices]), atol=1e3 * np.finfo(rdtype).eps
+        )
+        assert equal_l
+        assert comm.allreduce(equal_l, MPI.BAND)
+
+    x = ufl.SpatialCoordinate(msh)
+    V = fem.functionspace(msh, ("P", 1))
+    v = ufl.conj(ufl.TestFunction(V))
+
+    # Full domain
+    check_vertex_integral_against_sum(x[0] * v * ufl.dP, np.arange(num_vertices))
+
+    # Split domain into left half of vertices (1) and right half of vertices (2)
+    vertices_1 = mesh.locate_entities(msh, 0, lambda x: x[0] <= 0.5)
+    vertices_1 = vertices_1[vertices_1 < num_vertices]
+    vertices_2 = mesh.locate_entities(msh, 0, lambda x: x[0] > 0.5)
+    vertices_2 = vertices_2[vertices_2 < num_vertices]
+
+    tags = np.full(num_vertices, 1)
+    tags[vertices_2] = 2
+    vertices = np.arange(0, num_vertices, dtype=np.int32)
+    meshtags = mesh.meshtags(msh, 0, vertices, tags)
+
+    dP = ufl.Measure("dP", domain=msh, subdomain_data=meshtags)
+
+    check_vertex_integral_against_sum(x[0] * v * dP(1), vertices_1)
+    check_vertex_integral_against_sum(x[0] * v * dP(2), vertices_2)
+    check_vertex_integral_against_sum(x[0] * v * (dP(1) + dP(2)), np.arange(num_vertices))
+
+    V = fem.functionspace(msh, ("P", 1))
+    u = fem.Function(V, dtype=dtype)
+    u.x.array[:] = vertex_map.local_to_global(np.arange(num_vertices + vertex_map.num_ghosts))
+
+    check_vertex_integral_against_sum(u * x[0] * v * ufl.dP, np.arange(num_vertices), True)
+    check_vertex_integral_against_sum(u * x[0] * v * dP(1), vertices_1, True)
+    check_vertex_integral_against_sum(u * x[0] * v * dP(2), vertices_2, True)
+    check_vertex_integral_against_sum(u * x[0] * v * (dP(1) + dP(2)), np.arange(num_vertices), True)
```

Comment:
Could we add an additional check where the user manually packs the `(cell, local_vertex_index)` pair in the integration domain (could just be an extra line on one of the existing tests)`, to check that manual specification of integration entities also works.